components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -0,0 +1,408 @@
+
+    // Update T
+    const auto T_k = T(k);
+    T(k).set(at_top,                    T_k - fac*omega_int_kp1*delta_T_k);
+    T(k).set(at_bot,                    T_k - fac*omega_int_k*delta_T_km1);
+    T(k).set(not at_top and not at_bot, T_k - fac*(omega_int_kp1*delta_T_k + omega_int_k*delta_T_km1));
+    // Add thermal expansion term due to LS vertical advection
+    T(k) += dt*omega(k)*T_k*Rair/(Cpair*pmid(k));;
+
+    // Update Q
+    for (int iq=0; iq<n_q_tracers; ++iq) {
+      auto s_delta_tracer = Kokkos::subview(s_delta_Q, iq, Kokkos::ALL());
+      Pack delta_tracer_k, delta_tracer_km1;
+      ekat::index_and_shift<-1>(s_delta_tracer, range_pack_m1, delta_tracer_k, delta_tracer_km1);
+      Q(iq, k).set(at_top,                    Q(iq, k) - fac*omega_int_kp1*delta_tracer_k);
+      Q(iq, k).set(at_bot,                    Q(iq, k) - fac*omega_int_k*delta_tracer_km1);
+      Q(iq, k).set(not at_top and not at_bot, Q(iq, k) - fac*(omega_int_kp1*delta_tracer_k + omega_int_k*delta_tracer_km1));
+    }
+  });
+
+  // Release WS views
+  workspace.release_macro_block(delta_Q_slot, n_q_tracers);
+  workspace.release_many_contiguous<4>({&omega_int,  &delta_u,  &delta_v,  &delta_T});
+}
+
+// Apply large scale forcing for temperature and water vapor as provided by the IOP file
+KOKKOS_FUNCTION
+void HommeDynamics::
+advance_iop_forcing(const KT::MemberType& team,
+                         const int nlevs,
+                         const Real dt,
+                         const view_1d<const Pack>& divT,
+                         const view_1d<const Pack>& divq,
+                         const view_1d<Pack>& T,
+                         const view_1d<Pack>& qv)
+{
+  const auto nlev_packs = ekat::npack<Pack>(nlevs);
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&] (const int k) {
+    T(k) += dt*divT(k);
+    qv(k) += dt*divq(k);
+  });
+}
+
+void HommeDynamics::
+apply_iop_forcing(const Real dt)
+{
+  using ESU = ekat::ExeSpaceUtils<KT::ExeSpace>;
+
+  using EOS = Homme::EquationOfState;
+  using ElementOps = Homme::ElementOps;
+  using KV = Homme::KernelVariables;
+
+  using ColOps = ColumnOps<DefaultDevice, Real>;
+  using C = physics::Constants<Real>;
+  constexpr Real Rair = C::Rair;
+
+  // Homme objects
+  const auto& c = Homme::Context::singleton();
+  const auto& hvcoord = c.get<Homme::HybridVCoord>();
+  const auto& params = c.get<Homme::SimulationParams>();
+
+  // Dimensions
+  constexpr int NGP   = HOMMEXX_NP;
+  constexpr int NLEV  = HOMMEXX_NUM_LEV;
+  constexpr int NLEVI = HOMMEXX_NUM_LEV_P;
+  const auto nelem    = m_dyn_grid->get_num_local_dofs()/(NGP*NGP);
+  const auto total_levels = m_dyn_grid->get_num_vertical_levels();
+  const auto qsize = params.qsize;
+
+  // Sanity checks since we will be switching between ekat::Pack
+  // and Homme::Scalar view types
+  EKAT_ASSERT(NLEV  == ekat::npack<Pack>(total_levels));
```

Comment:
You could use `EKAT_ASSERT_MSG`, and add an explanation of what the root cause could be